### DISCUSSION TITLE
can I have toolchain config per sub-project?

### DISCUSSION THREAD

**Discussion by stsp** [mesonbuild/meson#14054] [Q&A]
Hi!

Currently my project consists of
separate subprojects. But the options
like --native-file and --cross-file seem
to be global. This leads to confusion,
for example meson forces me to
specify cpp compiler in cross-file,
even though the particular subproject
that needs cross-file, doesn't use c++.

Can I specify toolchain configs per
subproject?

**Comment by klokik** [comment]
No, you cannot. As you mentioned these are global. As I see it, meson tries to decouple environment configuration from subproject build definitions as much as possible.

Consider splitting C++ compiler definitions into a separate cross file an layer it on top of your primary one if you know that you will be building a subproject that requires it. https://mesonbuild.com/Machine-files.html#loading-multiple-machine-files

**Reply by stsp** [reply]
My primary one already has `cpp = 'clang++'`,
but its native, not cross file.
Meson requires `cpp` also in cross
file, likely because cross file does not
inherit from native. Given that cross
compilation in my case only needs
nasm (and no C or C++), having to write
C and C++ compilers into a cross file is
more than strange. I split the cross-compilation
into a sub-project which doesn't list c++,
but it is still mixed with the main project
so meson forces me to put C and C++
compilers into a cross file, when neither
is needed.

**Comment by stsp** [comment]
Maybe meson should support more
descriptive project declarations? Like:
`project('my_project', native: ['c', 'c++'], cross: ['nasm'])`?
